### ISSUE TITLE
Help with JSON Schema

### ISSUE THREAD

**Issue by matentzn** [OBOFoundry/OBOFoundry.github.io#1685]
I am trying to capture @cthoyt tests in JSON Schema.

I am wondering about this:
https://github.com/OBOFoundry/OBOFoundry.github.io/pull/1683/files?authenticity_token=f0OigYhnTMbqNI6Myxx704tAwrM%2B9YW%2B1llkaq5SE1KVgbMFSAOQKGIs0q8RJUXviyZLle%2Bh%2Bnemks3kKgIdyA%3D%3D&file-filters%5B%5D=.py#diff-d3729917c38df2205697e14a000d45f5f574d8c5204f05a1f2930111fb4a5b8fR85

How can I reflect this?

It seems to me that the tests themselves can _easily_ be added to the schema.

https://github.com/OBOFoundry/OBOFoundry.github.io/pull/1683/files?authenticity_token=f0OigYhnTMbqNI6Myxx704tAwrM%2B9YW%2B1llkaq5SE1KVgbMFSAOQKGIs0q8RJUXviyZLle%2Bh%2Bnemks3kKgIdyA%3D%3D&file-filters%5B%5D=.py#diff-d3729917c38df2205697e14a000d45f5f574d8c5204f05a1f2930111fb4a5b8fR93

But how do i tell JSON Schema that these conditions only apply to active ontologies?

**Comment by jamesaoverton** [comment]
I don't love JSON Schema, but we use it on many projects. One advantage of expressing the validation rules in a declarative data file rather than imperative code is that we can add additional fields (e.g. description) to the JSON Schema and reuse it for things like this context-aware YAML editor <https://editor.obolibrary.org/edit/registry/obi.md>. (I think this is a good idea, but the current editor breaks too often.)

As you point out, OBO has these "levels", and they don't fit well with JSON Schema. What we currently do is run the same JSON Schema on all the registry entries, but handle the `jsonschema.exceptions.ValidationError`s in different ways. There is a hideous conditional here that's doing most of that:

https://github.com/OBOFoundry/OBOFoundry.github.io/blob/master/util/validate-metadata.py#L151

This could definitely be implemented better, but I think the overall approach is fine.

TLDR: The JSON Schema only fully applies to active ontologies, with some exceptions specified by that ugly conditional.

**Comment by cmungall** [comment]
We can use schema composition to specify conditional logic

https://json-schema.org/understanding-json-schema/reference/combining.html

IMO we should use linkml for the schema, we can then specify the structure and the jsonld mappings all in one place - and still compile down to json-schema - this would be a longer term project though

**Comment by jamesaoverton** [comment]
I remember trying and failing to make this work with JSON Schema conditionals. We basically want all the rules to run on all the ontologies, but the project status determines what we count as an error vs a warning.

**Comment by cthoyt** [comment]
I have some ideas on this but I don't want to propose anything concrete until we sort out what the best way to add new / fix old code will be
